### PR TITLE
Fix useEffect not running when a dependency changes to nil

### LINKED ISSUE #31: Dependency changing to `nil` is not detected
For `useEffect`, when a dependency changes to `nil`, it doesn't re-run the function.
In this example, if the `visible` prop to `UserComponent` becomes `false`, it will not cancel the promise, even though the user id dependency changed to `nil`
```lua
local function useUsername(hooks, userId)
    local username, setUsername = hooks.useState(nil)
    hooks.useEffect(function()
        setUsername(nil)
        if userId then
            local promise = getNameFromId(userId):andThen(setUsername)
            return function()
                promise:cancel()
            end
        end
    end, {userId})
    return username
end

local function UserComponent(props, hooks)
    local username = useUsername(hooks, if props.visible then props.userId else nil)
    -- render user related stuff with username
end
```
I could work around this by passing `visible` as a dependency, but it makes the hook less reusable. It should detect that the dependency changed to `nil` and re-run the effect.

### PR DESCRIPTION
**Pull request by chriscerie**
Closes #31

### CODE REVIEW

**Comment by chriscerie** [comment]
Added relevant tests in #30

File: src/init.lua [hunk 1]
```diff
@@ -81,14 +81,22 @@ function Hooks.new(roact)
...
+							length += 1
 							if lastDependencies[dependencyIndex] ~= dependency then
 								anythingChanged = true
 								break
 							end
 						end
 
+						for _, _ in pairs(lastDependencies) do
+							length -= 1
+						end
```
Comment (Kampfkarren) [inline src/init.lua:95]:
```suggestion
						end

```

File: src/init.lua [hunk 2]
```diff
@@ -81,14 +81,22 @@ function Hooks.new(roact)
 					local lastDependencies = self.effectDependencies[index]
 					if lastDependencies ~= nil then
 						local anythingChanged = false
-
+						local length = 0
```
Comment (Kampfkarren) [inline src/init.lua:84]:
```suggestion
						local length = 0

```

File: src/init.lua [hunk 3]
```diff
@@ -81,14 +81,22 @@ function Hooks.new(roact)
...
 						for dependencyIndex, dependency in pairs(dependsOn) do
+							length += 1
```
Comment (Kampfkarren) [inline src/init.lua:86]:
```suggestion
							length += 1

```

File: src/init.lua [hunk 4]
```diff
@@ -81,14 +81,25 @@ function Hooks.new(roact)
...
+
 							if lastDependencies[dependencyIndex] ~= dependency then
 								anythingChanged = true
 								break
 							end
 						end
 
+						for _, _ in pairs(lastDependencies) do
```
Comment (Kampfkarren) [inline src/init.lua:95]:
```suggestion
						for _ in pairs(lastDependencies) do
```

**Pull request merged**